java/vector/src/main/codegen/templates/DenseUnionVector.java

```diff
@@ -0,0 +1,907 @@
+        }
+      }
+      to.setValueCount(length);
+
+      // transfer vector values
+      for (int i = 0; i < nextTypeId; i++) {
+        if (typeCounts[i] > 0 && typeStarts[i] != -1) {
+          internalTransferPairs[i].splitAndTransfer(typeStarts[i], typeCounts[i]);
+          to.childVectors[i] = internalTransferPairs[i].getTo();
+        }
+      }
+    }
+
+    @Override
+    public ValueVector getTo() {
+      return to;
+    }
+
+    @Override
+    public void copyValueSafe(int from, int to) {
+      this.to.copyFrom(from, to, DenseUnionVector.this);
+    }
+  }
+
+  @Override
+  public FieldReader getReader() {
+    if (reader == null) {
+      reader = new DenseUnionReader(this);
+    }
+    return reader;
+  }
+
+  public FieldWriter getWriter() {
+    if (writer == null) {
+      writer = new DenseUnionWriter(this);
+    }
+    return writer;
+  }
+
+  @Override
+  public int getBufferSize() {
+    return this.getBufferSizeFor(this.valueCount);
+  }
+
+  @Override
+  public int getBufferSizeFor(final int count) {
+    if (count == 0) {
+      return 0;
+    }
+    return count * TYPE_WIDTH + count * OFFSET_WIDTH + DataSizeRoundingUtil.divideBy8Ceil(count)
+        + internalStruct.getBufferSizeFor(count);
+  }
+
+  @Override
+  public ArrowBuf[] getBuffers(boolean clear) {
+    List<ArrowBuf> list = new java.util.ArrayList<>();
+    setReaderAndWriterIndex();
+    if (getBufferSize() != 0) {
+      list.add(validityBuffer);
+      list.add(typeBuffer);
+      list.add(offsetBuffer);
+      list.addAll(java.util.Arrays.asList(internalStruct.getBuffers(clear)));
+    }
+    if (clear) {
+      valueCount = 0;
+      validityBuffer.getReferenceManager().retain();
+      validityBuffer.close();
+      validityBuffer = allocator.getEmpty();
+      typeBuffer.getReferenceManager().retain();
+      typeBuffer.close();
+      typeBuffer = allocator.getEmpty();
+      offsetBuffer.getReferenceManager().retain();
+      offsetBuffer.close();
+      offsetBuffer = allocator.getEmpty();
+    }
+    return list.toArray(new ArrowBuf[list.size()]);
+  }
+
+  @Override
+  public Iterator<ValueVector> iterator() {
+    List<ValueVector> vectors = org.apache.arrow.util.Collections2.toList(internalStruct.iterator());
+    return vectors.iterator();
+  }
+
+  private ValueVector getVector(int index) {
+    byte typeId = typeBuffer.getByte(index * TYPE_WIDTH);
+    return getVectorByType(typeId);
+  }
+
+  public Object getObject(int index) {
+    if (isNull(index)) {
+      return null;
+    }
+    ValueVector vector = getVector(index);
+    if (vector != null) {
+      int offset = offsetBuffer.getInt(index * OFFSET_WIDTH);
+      return vector.getObject(offset);
+    }
+    return null;
+  }
+
+  public void get(int index, DenseUnionHolder holder) {
+    FieldReader reader = new DenseUnionReader(DenseUnionVector.this);
+    int offset = offsetBuffer.getInt(index * OFFSET_WIDTH);
+    reader.setPosition(offset);
+    holder.reader = reader;
```

Comment:
should there be a check to see if reader is already initialized?

Reply:
You mean we can reuse the reader here? I am not sure, because the reader is stateful and there can be more than one client for this vector.
The implementation here aligns with the sparse union vector.